### ISSUE TITLE
Adjacently tagged enum not deserialisable; ExpectedMap

### ISSUE THREAD

**Issue by happylinks** [zenlist/serde_dynamo#87]
I'm trying to deserialize my database which has a Map with a List as an enum, like background: ["Video", {url: "", duration:0}]
This should be supported by serde, but I'm now having a feeling it's not supported by serde_dynamo for some reason?

I found this issue where it's possible: https://github.com/serde-rs/serde/issues/2479

```rust
#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
pub struct VideoBackgroundPayload {
    pub url: String,
    pub duration: f32,
}

#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
#[serde(tag = "0", content = "1")]
pub enum Background {
    Test(String),
    Video(VideoBackgroundPayload),
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct User {
    id: String,
    background: Background,
}

// And deserialize them as strongly-typed data structures
let mut item = HashMap::new();

item.insert(
    "id".to_string(),
    AttributeValue::S("clnofe2cg0000ayeq6ryddrzw".to_string()),
);

let mut background_map = HashMap::new();
background_map.insert(
    "duration".to_string(),
    AttributeValue::N("7960".to_string()),
);
background_map.insert(
    "url".to_string(),
    AttributeValue::S("https://google.nl".to_string()),
);

let background = AttributeValue::M(background_map);
item.insert(
    "background".to_string(),
    AttributeValue::L(vec![
        AttributeValue::S("Video".to_string()),
        background,
    ]),
);

let user: Result<User, serde_dynamo::Error> = from_item(item.clone());
match user {
    Ok(user) => println!("user: {:#?}", user),
    Err(e) => println!("error: {:#?}", e),
}

Outputs:
error: Error(
    ExpectedMap,
)
```

deserialize_struct always calls deserialize_map and errors if it doesn't see a map as the value: https://github.com/zenlist/serde_dynamo/blob/2940408cccca71c32fdfd0e4ffdceb0d35b7be7a/src/de/deserializer.rs#L292

Let me know if I'm doing something wrong :)

**Comment by bryanburgers** [comment]
Does the same approach work when you use the equivalent in serde_json?

**Comment by happylinks** [comment]
I will properly test and confirm tomorrow when I'm at my desk again, but just quickly checking on my phone I see that it handles deserialize_struct in a way that also handles arrays: https://github.com/serde-rs/json/blob/4bc1eaa03a6160593575bc9bc60c94dba4cab1e3/src/de.rs#L1808

**Comment by happylinks** [comment]
Can confirm that this works:
```rust
let user = r#"
    {
        "id": "clnofe2cg0000ayeq6ryddrzw",
        "background": ["Video", {"duration": 7960, "url": "https://google.nl"}]
    }"#;

// Parse the string of data into serde_json::Value.
let user: Result<User, serde_json::Error> = serde_json::from_str(user);
match user {
    Ok(user) => println!("user: {:#?}", user),
    Err(e) => println!("error: {:#?}", e),
}
```
```
user: User {
    id: "clnofe2cg0000ayeq6ryddrzw",
    background: Video(
        VideoBackgroundPayload {
            url: "https://google.nl",
            duration: 7960.0,
        },
    ),
}
```

**Comment by bryanburgers** [comment]
I didn't even realize this was a thing!

**Comment by bryanburgers** [comment]
Thanks for reporting this! Fixed in 4.2.10.

**Comment by happylinks** [comment]
Ah thanks so much for the fast fix! Awesome 🎉
